node/src/components/contract_runtime/rewards.rs

```diff
@@ -0,0 +1,413 @@
+
+        let result = collect_past_blocks_batched(effect_builder, range_to_fetch.clone()).await;
+
+        tracing::info!(
+            era_id = %era_id.value(),
+            range_requested = ?range_to_fetch,
+            num_fetched_blocks = %result.iter().flatten().count(),
+            num_requested_blocks = %result.len(),
+            "blocks fetched",
+        );
+
+        result
+    };
+
+    let eras_info = {
+        let mut cited_blocks = cited_blocks.iter().flatten();
+
+        // We take the first block, then every switch block to get all the needed hashes to fetch
+        // needed eras. We take the first block, because we need it for the first era.
+        // If the first block is itself a switch block, that's fine, because we fetch one block more
+        // to handle this case.
+
+        let oldest_block = cited_blocks.next();
+
+        // If the oldest block is genesis, we add the validator information for genesis (era 0) from
+        // era 1, because it's the same:
+        let oldest_block_is_genesis = oldest_block.map_or(false, |block| block.is_genesis());
+
+        let hashes_and_eras: Vec<_> = oldest_block
+            .into_iter()
+            .chain(cited_blocks.filter(|&block| block.is_switch()))
+            .map(|block| {
+                (
+                    // The validator info for a switch block is the one from the next era:
+                    if block.is_switch() {
+                        block.era_id().successor()
+                    } else {
+                        block.era_id()
+                    },
+                    *block.hash(),
+                )
+            })
+            .collect();
+
+        let num_eras_to_fetch = hashes_and_eras.len() + usize::from(oldest_block_is_genesis);
+
+        let mut result = ErasInfo::populate(
+            effect_builder,
+            chainspec.protocol_version(),
+            hashes_and_eras,
+        )
+        .await
+        .map_err(RewardsError::PopulateError)?;
+
+        if oldest_block_is_genesis {
+            result.copy_genesis_info_from_era_1()?;
+        }
+
+        tracing::info!(
+            era_id = %era_id.value(),
+            %num_eras_to_fetch,
+            eras_fetched = ?result.era_ids(),
+        );
+
+        result
+    };
+
+    let collection_proportion =
+        to_ratio_u512(chainspec.core_config.collection_rewards_proportion());
+    let contribution_proportion =
+        to_ratio_u512(chainspec.core_config.contribution_rewards_proportion());
+
+    // Reward for producing a block from this era:
+    let production_reward = to_ratio_u512(chainspec.core_config.production_rewards_proportion())
+        .checked_mul(&eras_info.reward(era_id)?)
+        .ok_or(RewardsError::ArithmeticOverflow)?;
+
+    // Collect all rewards as a ratio:
+    let full_reward_for_validators = {
+        let mut result = BTreeMap::new();
+
+        let blocks_from_current_era = cited_blocks.iter().filter_map(|maybe_block| {
```

Comment:
It _seems_ that we do reward the lagging signatures here (see line 332), so why do we first fetch than discard blocks from the preceding era? Just to obtain the era id and the switch block root hash?

Reply:
We don't discard them (that's why it's a borrowed `iter` and not `into_iter`. I keep them around because they may end up being one of the reported signing blocks.